### PR TITLE
Composite: stabilize APIs

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Requires #63564, #64399, #64397, #64450, #64493 and #64723

Part of https://github.com/WordPress/gutenberg/issues/58850

Export `Composite` as a stable API and refactor usages of the component from using the private APIs to using the public ones..

Also, remove the private APIs since the component is now public.

## Why?

This is one of the final steps in making the `Composite` component stable and available as a public API of `@wordpress/components`.

### Main changes

- Export the `Composite` component from the `@wordpress/components` package as stable APIs
- Remove the equivalent private API
- Replace all imports via private APIs with public APIs;
- Update component naming as necessary, using the new "overloaded" naming convention adopted for compound components;

## Testing Instructions

Given that the project builds and tests pass, most of the changes don't really need testing, since the underlying composite components are the exact same (we only changed the way they are imported).

The only real testing that should be done is around the changes where the unnecessary `store` and `onFocusVisible` props were removed.

### A new stable `Composite` component

Following the rewrite of the `__unstableComposite`  (& related) components from `ariakit` to `reakit` in the WordPress 6.5 release, the 6.7 release introduces a new, stable version for the `Composite` component.

The new `Composite` component offers a more complete and versatile way to implement low-level composite widgets, and its APIs are more consistent with the rest of the `@wordpress/components` package.

```tsx
import { Composite } from '@wordpress/components';

<Composite>
  <Composite.Item>Item one</Composite.Item>
  <Composite.Item>Item two</Composite.Item>
  <Composite.Item>Item three</Composite.Item>
</Composite>
```

For more information of the component's APIs and other usage examples, please refer to [the Storybook page](https://wordpress.github.io/gutenberg/?path=/docs/components-composite--docs).

Finally, the `__unstableComposite`  (& related) components have been marked as deprecated.

### CODE REVIEW

File: packages/components/src/index.ts
Comment (ciampo) [inline packages/components/src/index.ts]: I had actually forgotten to export `useCompositeStore` in previous PRs 😅

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (ciampo) [inline packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: Hey @ntsekouras could you help me understand if we can actually remove the `onFocusVisible` callback here? That's because in theory, from what I understand from [the docs](https://ariakit.org/reference/composite#onfocusvisible), it shouldn't work without the `focusable` prop.

Does it currently work? If not, should we remove it altogether, or try to make it work by also passing the `focusable` prop?

Thank you!

File: packages/dataviews/src/dataviews-layouts/list/index.tsx
Comment (oandregal) [inline packages/dataviews/src/dataviews-layouts/list/index.tsx:51]: The changes for this component seem to be one-to-one according to the names. However, something is different. The keyboard navigation for the list layout is broken in this PR (see also e2e test). Note how the dropdown action menu is not reachable via right arrow:

https://github.com/user-attachments/assets/d726753c-6c5a-4e59-8eb5-37ed65e20608

References, in case it helps:

- implementation https://github.com/WordPress/gutenberg/pull/60805
- reported issues with ariakit to make 2d keyboard navigation work https://github.com/ariakit/ariakit/issues/3768
- e2e test  https://github.com/WordPress/gutenberg/pull/61648

Comment (ciampo) [reply packages/dataviews/src/dataviews-layouts/list/index.tsx:51]: Could it be because I also removed passing the `store` prop to `Composite.Item` twice in this same file ? If you add it back, does it solve the issue for you?

Comment (ciampo) [reply packages/dataviews/src/dataviews-layouts/list/index.tsx:51]: Update: passing explicitly the `store` to `Composite.Item` seems to fix it, although that's an unexpected behaviour from my side that I'd like to investigate further. I'll post a new inline comment on that specific line.

File: packages/dataviews/src/dataviews-layouts/list/index.tsx
Comment (ciampo) [inline packages/dataviews/src/dataviews-layouts/list/index.tsx]: For now, we have to continue passing explicitly a `store` object to this `Composite.Item` instance — without it, the the composite item is not included in the parent composite widget (see [conversation](https://github.com/WordPress/gutenberg/pull/63569#discussion_r1713359138)).

After some debugging, it looks like the explicit `store` override is necessary because otherwise this instance of `Composite.Item` gets the `DropdownMenu`'s store instead of the parent `Composite` store when reading the store from context.

@diegohaz , is this the expected behaviour (ie. passing `store` explicitly), or is this due to how `DropdownMenu` is implemented ([code](https://github.com/WordPress/gutenberg/blob/039bd9081abeb7ce93d7a57e56ca0abfbc6b7937/packages/components/src/dropdown-menu-v2/index.tsx#L309)) ? I could replicate it also with vanilla Ariakit ([stackblitz](https://stackblitz.com/edit/grzwwc?file=menu%2Findex.tsx)), but I wonder if I'm missing something.

Thank you!

File: packages/components/src/composite/stories/index.story.tsx
Comment (tyxla) [inline packages/components/src/composite/stories/index.story.tsx]: According to the Ariakit docs, the `accessibleWhenDisabled` prop is also available and supported for the `Composite` element. Should we add it there as well?

File: packages/components/src/composite/types.ts
Comment (tyxla) [inline packages/components/src/composite/types.ts]: Same as above, wrt providing the prop type for `Composite`.

File: packages/components/src/index.ts
Comment (tyxla) [reply packages/components/src/index.ts]: Good timing to find out about it 😅

File: packages/components/src/private-apis.ts
Comment (tyxla) [inline packages/components/src/private-apis.ts]: Glad to see `@wordpress/components` private APIs are down with over 25% 👏

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (tyxla) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: From what reading I've done, I think adding `focusable` should do it. It doesn't seem to work in `trunk` for me in my testing right now.

**Review by tyxla — APPROVED** [review]
Aside from `onFocusVisible` + `focusable`, this looks great to me 👍 

Nice work @ciampo 🚀 

Let's make sure to add a CHANGELOG entry before 🚢 .

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (tyxla) [inline packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: 👍

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (ciampo) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: In my experience, testing those lines of code was tricky because of some different behaviors based on the number of items and the operators' lenght, but I wasn't able to notice regressions. What regressions have you noticed?

Since `onFocusVisible` is necessary, I went ahead and added it to `Composite`'s props, alongside a few focus-related props (`focusable`, `disabled`, `accessibleWhenDisabled`, `onFocusVisible`).

--

Specifically to why `onFocusVisible` is being triggered even without an explicit `focusable` prop, I'm going to continue the investigation in [a separate conversation](https://github.com/WordPress/gutenberg/pull/63569#discussion_r1714043891) to get better clarity.

File: packages/components/src/composite/stories/index.story.tsx
Comment (ciampo) [reply packages/components/src/composite/stories/index.story.tsx]: Added, together with other props, as explained in https://github.com/WordPress/gutenberg/pull/63569#discussion_r1714041967

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (ciampo) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: @diegohaz can you confirm that my conclusions are correct? Is this the expected behavior? Should we pass `focusable` to `Composite` to future-proof this behavior?

**Comment by ciampo** [comment]
> this looks great to me 👍

Thank you! I'd like to get clarity both on the [`store`](https://github.com/WordPress/gutenberg/pull/63569#discussion_r1713930889) and [`onFocusVisible`](https://github.com/WordPress/gutenberg/pull/63569#discussion_r1714043891) prop behaviors before merging, assuming we're going to get a reply soon.

> Let's make sure to add a CHANGELOG entry before 🚢 .

I'm also considering opening a separate PR with the changes to `Composite`, so that we keep this PR focused on the rest of the repository (#64450)

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (tyxla) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: Actually, it works as expected, both in trunk and here:

https://github.com/user-attachments/assets/07b735ae-0167-4935-8d42-8a75d4342a5c

so we should be good 👍

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (diegohaz) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: The description makes sense.

> Should we pass `focusable` to `Composite` to future-proof this behavior?

Not all composite containers are focusable, though. For example, a toolbar container is not focusable and therefore doesn't trigger `onFocusVisible`.

File: packages/dataviews/src/dataviews-layouts/list/index.tsx
Comment (diegohaz) [reply packages/dataviews/src/dataviews-layouts/list/index.tsx]: That's "expected" because the Ariakit Menu component also provides a Composite context. I acknowledge this is desirable in some cases but cumbersome in others, so this may change in the future.

Related issue: [https://github.com/ariakit/ariakit/issues/3754](https://github.com/ariakit/ariakit/issues/3754)

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (diegohaz) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: You can think of `focusable`/`onFocusVisible` as similar to `draggable`/`onDragStart`.

**Comment by tyxla** [comment]
#64450 looks good 👍 Let's merge it, and this one will be much smaller. I'm happy to re-review once it's ready.

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (ntsekouras) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: Sorry for the late reply. It seems you figured it out already 😄 . The conditional rendering of different components and what designers wanted led to this and similar nuances like [this one](https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/src/components/dataviews-filters/search-widget.tsx#L90).

File: packages/dataviews/src/dataviews-layouts/list/index.tsx
Comment (ciampo) [reply packages/dataviews/src/dataviews-layouts/list/index.tsx]: Ah, that explains it! I didn't consider that different components could internally use another composite store instance.

I think that the best approach for us here is to use a custom `Composite` context to make sure that all sub-components consume the correct store (it's the same approach that we adopted in [`Tabs`](https://github.com/WordPress/gutenberg/blob/039bd9081abeb7ce93d7a57e56ca0abfbc6b7937/packages/components/src/tabs/index.tsx#L198) and [`DropdownMenuV2`](https://github.com/WordPress/gutenberg/blob/039bd9081abeb7ce93d7a57e56ca0abfbc6b7937/packages/components/src/dropdown-menu-v2/index.tsx#L331))

I tested those changes locally and they work well, and I opened a separate PR to land them: #64493

I already removed the explicit `store` prop from `Composite.Item`, and will rebase on top of `trunk` once #64493 is merged. We should be ready to land this PR by then.

File: packages/dataviews/src/components/dataviews-filters/search-widget.tsx
Comment (ciampo) [reply packages/dataviews/src/components/dataviews-filters/search-widget.tsx]: > > Should we pass `focusable` to `Composite` to future-proof this behavior?
> 
> Not all composite containers are focusable, though.

That makes sense! Although my question was specific to this instance of `Composite`. My inline code comment is hopefully enough to explain why `onFocusVisible` is being called despite the `focusable` prop is not set on the component.

File: packages/dataviews/src/dataviews-layouts/list/index.tsx
Comment (tyxla) [reply packages/dataviews/src/dataviews-layouts/list/index.tsx]: An internal context makes sense for passing down the store.  👍 

#64493 looks good! ✅ 

Noting that there appears to be one more instance where we explicitly pass `store` to a `Composite.Item`:

https://github.com/WordPress/gutenberg/blob/2c62c6aa2f30060b3acd40b225aebe4295ad0bf4/packages/block-directory/src/components/downloadable-block-list-item/index.js#L117

Might make sense to address it here before landing.

**Comment by ciampo** [comment]
Update: with #64493 merged, this PR should be good to be merged too. I'll wait for CI, run some extra manual tests, and merge if everything looks good.

**Comment by tyxla** [comment]
Were you also planning to address this before merging:

https://github.com/WordPress/gutenberg/pull/63569#discussion_r1716812603

File: packages/dataviews/src/dataviews-layouts/list/index.tsx
Comment (ciampo) [reply packages/dataviews/src/dataviews-layouts/list/index.tsx]: I had somewhat missed that instance (and your comment). Thank you for flagging again

**Review by tyxla — CHANGES_REQUESTED** [review]
Let's hold off merging this for now until we settle how we'll deal with some of the public API considerations, see #64510 and #63704.

**Comment by ciampo** [comment]
@WordPress/gutenberg-components This PR is now ready for a new round of review, after addressing all the necessary API changes in #64723 & related.

File: packages/components/src/private-apis.ts
Comment (tyxla) [inline packages/components/src/private-apis.ts:21]: So cool to see `@wordpress/components` private APIs going down by half 👏

**Review by tyxla — APPROVED** [review]
Did some testing in the editor and in Storybook, everything works well!

Let's go! 🚀 

Great work addressing it incrementally @ciampo!

**Pull request merged**
